frontend_lib/src/helper.js

```diff
@@ -927,6 +926,213 @@ export const handleClickCopyLink = (contentId) => {
   } else navigator.clipboard.writeText(link)
 }
 
+// /////////////////////////////////////////////////////////////////////////////
+// NOTE - MP - 2022-12-02 - MENTION SECTION
+// /////////////////////////////////////////////////////////////////////////////
+
+/**
+ * Search mention in a string
+ * @param {String} text The text to search mentions in
+ * @returns {List[String]} List of mentions found
+ * Example:
+ * - Input: `<p>Test @Jhon</p>`
+ * - Output: `['@Jhon']`
+ */
+export const searchMention = (text) => {
+  // Regex explanation: https://regex101.com/r/hHosBa/10
+  // Match (@XXX part): '@XXX', ' @XXX ', '@XXX-', ':@XXX:', '(@XXX)', '!@XXX!', ...
+  // Don't match: 'XXX@XXX', '@<span>XXX</span>'
+  const mentionRegex = /(?<=^|\s|\W)@([a-zA-Z0-9_-]+)\b/g
+  const mentionList = text.match(mentionRegex)
+  return mentionList || []
+}
+
+/**
+ * Replace not formatted mention with html mention element
+ * @param {List[role]} roleList List of role that can be mentionned
+ * @param {List[user]} userList List of user that can be mentionned
+ * @param {String} html Current content of the editor
+ * @returns {{html: String, invalidMentionList: List[String]}} Correctly formatted html content
+ * Example:
+ * - Input: `<p>Test @Jhon</p>`
+ * - Output:\
+ * {\
+ *   html: `<p>Test <html-mention userid="151"/></p>`;\
+ *   invalidMentionList: [];\
+ * }
+ */
+export const searchMentionAndPlaceBalise = (rolelist, userList, html) => {
+  const mentionList = searchMention(html)
+  const invalidMentionList = []
+
+  let newHtml = html
+
+  mentionList.forEach(mention => {
+    const mentionWithoutAt = mention.slice(1)
+    const role = rolelist.find(r => r.slug === mentionWithoutAt)
+    const user = userList.find(u => u.username === mentionWithoutAt)
+    if (role || user) {
+      const mentionBalise = `<html-mention ${
+        role ? 'roleid' : 'userid'
+      }="${
+        role ? role.id : user.id
+      }"></html-mention>`
+      newHtml = newHtml.replace(mention, mentionBalise)
+    } else {
+      invalidMentionList.push(mention)
+    }
+  })
+
+  return { html: newHtml, invalidMentionList }
+}
+
+/**
+ * Replace the given HTML string containing the mention with the role slug
+ * @param {Array[object]} roleList List of roles
+ * @param {String} html Current html text with balise mention
+ * @returns Html text without mention balise
+ * Example:
+ * - Input: `<p>Test <html-mention roleid="0"><\html-mention></p>`
+ * - Output: `<p>Test @all</p>`
+ */
+const replaceHTMLElementWithMentionRole = (roleList, html) => {
+  const mentionRegex = /<html-mention roleid="(\d+)"><\/html-mention>/g
```

Comment:
I think ``rolelevel`` is more appropriate here, since it's compared to level (in backend to send the notification and should be compared to the level in the frontend to know if the mention is for us)

However, in the frontend everywhere we reference to the ``rolelevel`` we use the ``id`` of the ``role``... That's why I used ``roleid``